ext/pdo_pgsql/pgsql_driver.c

```diff
@@ -661,34 +688,35 @@ void pgsqlCopyFromArray_internal(INTERNAL_FUNCTION_PARAMETERS)
 
 	if (status == PGRES_COPY_IN && pgsql_result) {
 		int command_failed = 0;
-		size_t buffer_len = 0;
 		zval *tmp;
+		zend_object_iterator *iter;
 
 		PQclear(pgsql_result);
-		ZEND_HASH_FOREACH_VAL(Z_ARRVAL_P(pg_rows), tmp) {
-			size_t query_len;
-			if (!try_convert_to_string(tmp)) {
-				efree(query);
+
+		if (Z_TYPE_P(pg_rows) == IS_ARRAY) {
+			ZEND_HASH_FOREACH_VAL(Z_ARRVAL_P(pg_rows), tmp) {
+				if (!_pdo_pgsql_send_copy_data(H, tmp)) {
+					efree(query);
+					RETURN_THROWS();
+				}
+			} ZEND_HASH_FOREACH_END();
+		} else {
+			iter = Z_OBJ_P(pg_rows)->ce->get_iterator(Z_OBJCE_P(pg_rows), pg_rows, 0);
+			if (iter == NULL || EG(exception)) {
 				RETURN_THROWS();
 			}
 
-			if (buffer_len < Z_STRLEN_P(tmp)) {
-				buffer_len = Z_STRLEN_P(tmp);
-				query = erealloc(query, buffer_len + 2); /* room for \n\0 */
-			}
-			query_len = Z_STRLEN_P(tmp);
-			memcpy(query, Z_STRVAL_P(tmp), query_len);
-			if (query[query_len - 1] != '\n') {
-				query[query_len++] = '\n';
-			}
-			query[query_len] = '\0';
-			if (PQputCopyData(H->server, query, query_len) != 1) {
-				efree(query);
-				pdo_pgsql_error(dbh, PGRES_FATAL_ERROR, NULL);
-				PDO_HANDLE_DBH_ERR();
-				RETURN_FALSE;
+			for (; iter->funcs->valid(iter) == SUCCESS && EG(exception) == NULL; iter->funcs->move_forward(iter)) {
+				tmp = iter->funcs->get_current_data(iter);
+				if (!_pdo_pgsql_send_copy_data(H, tmp)) {
+					efree(query);
+					zend_iterator_dtor(iter);
+					RETURN_THROWS();
+				}
 			}
-		} ZEND_HASH_FOREACH_END();
+			zend_iterator_dtor(iter);
+		}
+
 		if (query) {
```

Comment:
same remark as above anyway by the look at the new code, it s not happening anyway.

Reply:
Thank you, I overlooked that.
I deleted all 3 places with af02672.